### PR TITLE
[Merged by Bors] - feat(Data.Set.Ncard): enat set cardinality

### PR DESCRIPTION
**Pull request by apnelson1**
This PR is a second attempt at defining the cardinality of a set as extended natural number, with a function `encard`. The implementation involves a refactor, where the existing`ncard` function is redefined in terms of `encard`. This shortens a lot of proofs, reduces reliance on the `Finset` API, and allows for a potential future refactor where `ncard` is removed if it is decided to be redundant.  

---

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

File: Mathlib/Data/Set/Ncard.lean
Comment (j-loreaux) [inline Mathlib/Data/Set/Ncard.lean:370]:
```suggestion
theorem _root_.Function.Injective.encard_image (hf : f.Injective) (s : Set α) : (f '' s).encard = s.encard :=
```

File: Mathlib/Data/Set/Ncard.lean
Comment (j-loreaux) [inline Mathlib/Data/Set/Ncard.lean:373]:
```suggestion
theorem _root_.Function.Embedding.encard_le (e : s ↪ t) : s.encard ≤ t.encard := by
```

File: Mathlib/Data/Set/Ncard.lean
Comment (j-loreaux) [inline Mathlib/Data/Set/Ncard.lean:393]:
```suggestion
theorem _root_.Function.Injective.encard_preimage_of_subset_range (hf : f.Injective) (ht : t ⊆ range f) :
```

File: Mathlib/Data/Set/Ncard.lean
Comment (j-loreaux) [inline Mathlib/Data/Set/Ncard.lean:398]: `hf : MapsTo f s t`, and you probably want to rename this to take advantage of dot notation.

File: Mathlib/Data/Set/Ncard.lean
Comment (j-loreaux) [inline Mathlib/Data/Set/Ncard.lean:64]: Here and everywhere else: I think our plan is to leave the `#align`s around for a while (even if they are no longer perfect), so please put them back.

File: Mathlib/Data/Set/Ncard.lean
Comment (j-loreaux) [inline Mathlib/Data/Set/Ncard.lean:624]:
```suggestion
  congr_arg ENat.toNat <| encard_exchange ha hb
```

**Review by j-loreaux — COMMENTED** [review]
This looks pretty nice thanks! I left a bunch of suggestions, but they are minor. Thanks for refactoring `Set.ncard` to use `Set.encard`.

We may want to rename the file (in a different PR probably) to `Data.Set.Card` after this is merged.

**Comment by apnelson1** [comment]
> This looks pretty nice thanks! I left a bunch of suggestions, but they are minor. Thanks for refactoring `Set.ncard` to use `Set.encard`.
> 
> We may want to rename the file (in a different PR probably) to `Data.Set.Card` after this is merged.

Thanks for the comments. I'm on board with the rename - I'll PR it once this one is through.

**Comment by j-loreaux** [comment]
@apnelson1 just FYI, I think you may have missed a bunch of the comments (GitHub automatically folds long sections of the PR/comment history and you have to click to unfold it).

File: Mathlib/Data/Set/Ncard.lean
Comment (apnelson1) [reply Mathlib/Data/Set/Ncard.lean:100]: `simp` already reduces the LHS to `¬(s = ∅)` so the linter dislikes this.

**Comment by apnelson1** [comment]
> @apnelson1 just FYI, I think you may have missed a bunch of the comments (GitHub automatically folds long sections of the PR/comment history and you have to click to unfold it).

Oops/Thanks! - I think they're all addressed now.

File: Mathlib/Data/Set/Ncard.lean
Comment (j-loreaux) [inline Mathlib/Data/Set/Ncard.lean:389]:
```suggestion
  rw [←encard_univ_coe, ←e.injective.encard_image, ←Subtype.coe_injective.encard_image]
```

File: Mathlib/Data/Set/Ncard.lean
Comment (j-loreaux) [inline Mathlib/Data/Set/Ncard.lean:445]:
```suggestion
    (h.symm.trans hinj.encard_image.symm).le]
```

**Review by j-loreaux — APPROVED** [review]
Two sets of extra parentheses left, but I'm delegating this to you. Please don't merge before CI finishes.

**Comment by j-loreaux** [comment]
bors d+

**Pull request closed without merging**
